x/mongo/driver/topology/topology.go

```diff
@@ -390,39 +426,60 @@ func wrapServerSelectionError(err error, t *Topology) error {
 	return fmt.Errorf("server selection error: %v, current topology: { %s }", err, t.String())
 }
 
-// selectServer is the core piece of server selection. It handles getting
-// topology descriptions and running sever selection on those descriptions.
-func (t *Topology) selectServer(ctx context.Context, subscriptionCh <-chan description.Topology, ss description.ServerSelector, timeoutCh <-chan time.Time) ([]description.Server, error) {
+// selectServerFromSubscription loops until a topology description is available for server selection. It returns
+// when the given context expires, server selection timeout is reached, or a description containing a selectable
+// server is available.
+func (t *Topology) selectServerFromSubscription(ctx context.Context, subscriptionCh <-chan description.Topology,
+	selectionState serverSelectionState) ([]description.Server, error) {
+
 	var current description.Topology
 	for {
 		select {
 		case <-ctx.Done():
 			return nil, ctx.Err()
-		case <-timeoutCh:
+		case <-selectionState.timeoutChan:
 			return nil, wrapServerSelectionError(ErrServerSelectionTimeout, t)
 		case current = <-subscriptionCh:
 		}
 
-		var allowed []description.Server
-		for _, s := range current.Servers {
-			if s.Kind != description.Unknown {
-				allowed = append(allowed, s)
-			}
-		}
-
-		suitable, err := ss.SelectServer(current, allowed)
+		suitable, err := t.selectServerFromDescription(ctx, current, selectionState)
 		if err != nil {
-			return nil, wrapServerSelectionError(err, t)
+			return nil, err
 		}
 
 		if len(suitable) > 0 {
 			return suitable, nil
 		}
-
 		t.RequestImmediateCheck()
 	}
 }
 
+// selectServerFromDescription process the given topology description and returns a slice of suitable servers.
+func (t *Topology) selectServerFromDescription(ctx context.Context, desc description.Topology,
+	selectionState serverSelectionState) ([]description.Server, error) {
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-selectionState.timeoutChan:
```

Comment:
I raised this in the ticket as well, but it seems like checking this here might mean that if you set ServerSelectionTimeout low enough you might spuriously fail to select e.g. a primary (e.g. if you context switched just before this point, or ran GC)?